### PR TITLE
[grafana] add value to make extraConfigmapMounts and extraSecretMounts optional

### PR DESCRIPTION
**Pull request by tibuntu**
Hi there, we are using that feature a lot but recently ran into some kind of race condition because a referenced ConfigMap wasn't existing yet. I noticed that there is now also an alerting sidecar which will probably improve our situation, but I still wanted to contribute this small changes which gives one the option to make the extraConfigmapMounts optional.

### CODE REVIEW

**Comment by jkroepke** [comment]
Hi, could you add this feature to extraSecretMounts as well?

**Comment by tibuntu** [comment]
> Hi, could you add this feature to extraSecretMounts as well?

That makes sense, done!

**Comment by tibuntu** [comment]
Just noticed that I used different default values for the examples. What would you prefer?

File: charts/grafana/templates/_pod.tpl [hunk 1]
```diff
@@ -1156,6 +1156,7 @@ volumes:
   - name: {{ tpl .name $root }}
     configMap:
       name: {{ tpl .configMap $root }}
+      optional: {{ .optional }}
```
Comment (jkroepke) [inline charts/grafana/templates/_pod.tpl:1159]:
```suggestion
      {{- with .optional }}
      optional: {{ .optional }}
      {{- end }}
```

File: charts/grafana/templates/_pod.tpl [hunk 2]
```diff
@@ -1261,6 +1262,7 @@ volumes:
     secret:
       secretName: {{ .secretName }}
       defaultMode: {{ .defaultMode }}
+      optional: {{ .optional }}
```
Comment (jkroepke) [inline charts/grafana/templates/_pod.tpl:1265]:
```suggestion
      {{- with .optional }}
      optional: {{ .optional }}
      {{- end }}
```

**Comment by jkroepke** [comment]
optional: false is the Kubernetes default, lets use that.

I added a suggestion, if optional is not present at values, the key will be omit in the template, too.

**Review by jkroepke — APPROVED** [review]

**Review by zanhsieh — APPROVED** [review]

**Comment by makkes** [comment]
If the ConfigMap doesn't exist at the time the Pod is created, it won't be mounted into the container. Kubernetes itself doesn't restart the container as soon as the CM is created so what is the expected behaviour here? Should users delete the Pod after the CM is created? I suppose this can lead to a lot of confusion.

**Comment by jkroepke** [comment]
> Should users delete the Pod after the CM is created? 

Yes, that expected from Kubernetes point of view. 

> I suppose this can lead to a lot of confusion.

Yes, avoid using optional flag would be the best possible option here.

Not sure, if you can use https://github.com/stakater/Reloader to automate the restart.


----

The general recommendation from an Kuberneter guy with much experience: Avoid optional unless you know the consequences.

**Comment by makkes** [comment]
> The general recommendation from me with much experience: Avoid optional unless you know the consequences.

I agree and that's what my comment was all about that the behaviour is documented in the default `values.yaml` file accompanying this chart so that users don't expect the chart to do any magic around this but will rather just pass the flag to the mount. Not sure if this is a usual practice in charts, though.

**Pull request merged**